src/Http/Http/src/Features/ResponseBodyPipeFeature.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    public class ResponseBodyPipeFeature : IResponseBodyPipeFeature
+    {
+        private StreamPipeWriter _internalPipeWriter;
+        private PipeWriter _userSetPipeWriter;
+        private HttpContext _context;
+
+        public ResponseBodyPipeFeature(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            _context = context;
+        }
+
+        public PipeWriter ResponseBodyPipe
+        {
+            get
+            {
+                if (_userSetPipeWriter != null)
+                {
+                    return _userSetPipeWriter;
+                }
+
+                if (_internalPipeWriter == null)
+                {
+                    var streamPipeWriter = new StreamPipeWriter(_context.Response.Body);
+                    _internalPipeWriter = streamPipeWriter;
+                }
+                else if (!object.ReferenceEquals(_internalPipeWriter.InnerStream, _context.Response.Body))
+                {
+                    _internalPipeWriter = new StreamPipeWriter(_context.Response.Body);
+                    _context.Response.RegisterForDispose(_internalPipeWriter);
```

Comment:
We have to call RegisterForDispose here because by default the DefaultHttpContext doesn't call Uninitialize on the Request and Response features. https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http/src/DefaultHttpContext.cs#L197.

I'm still thinking about how to write a test for this.